File: Grid.cs
using System;
using SFML.Graphics;

namespace GOL{
    public class Grid {
        public int GridSize{get;}

        bool[] grid;

        public Image GridImage{get;private set;}
        public Texture GridTexture{get;private set;}

        float threashold;

        public Grid(int gridSize, float threashold){
            GridSize=gridSize;
            GridImage=new Image((uint)gridSize, (uint)gridSize);
            grid=new bool[gridSize*gridSize];
            GridTexture=new Texture(GridImage);
            this.threashold=threashold;
            Reset();
        }

        public void Reset(){
            Random rand=new Random();
            for(int i=0;i<grid.Length;i++){
                grid[i]=rand.Next(101)/100f>threashold;
            }
        }

        public void Update(){
            bool[] tempGrid=(bool[])grid.Clone();

            for(int y=0;y<GridSize;y++){
                for(int x=0;x<GridSize;x++){
                    int res=0;

                    int xx0=x-1;
                    int yy0=y-1;
                    int xx1=x+1;
                    int yy1=y+1;

                    if(x!=0)
                        res+=tempGrid[xx0+GridSize*y]?1:0;

                    if(y!=0)
                        res+=tempGrid[x+GridSize*yy0]?1:0;

                    if(x!=GridSize-1)
                        res+=tempGrid[xx1+GridSize*y]?1:0;

                    if(y!=GridSize-1)
                        res+=tempGrid[x+GridSize*yy1]?1:0;

                    if(x!=0 && y!=0)
                        res+=tempGrid[xx0+GridSize*yy0]?1:0;

                    if(x!=GridSize-1 && y!=0)
                        res+=tempGrid[xx1+GridSize*yy0]?1:0;

                    if(x!=0 && y!=GridSize-1)
                        res+=tempGrid[xx0+GridSize*yy1]?1:0;

                    if(x!=GridSize-1 && y!=GridSize-1)
                        res+=tempGrid[xx1+GridSize*yy1]?1:0;

                    if(!(tempGrid[x+GridSize*y]&&(res==3||res==2)))
                        grid[x+GridSize*y]=false;
                    if(!tempGrid[x+GridSize*y] && res==3)
                        grid[x+GridSize*y]=true;
                }
            }

            for(uint x=0;x<GridImage.Size.X;x++){
                for(uint y=0;y<GridImage.Size.Y;y++){
                    GridImage.SetPixel(x,y, grid[y*GridImage.Size.X+x]?Color.White:Color.Black);
                }
            }
            GridTexture.Update(GridImage);
        }
    }
}

File: src/grid.cpp
#include "grid.h"
#include <iostream>

Grid::Grid(int gridSize, float threashold) {
    m_gridSize = gridSize;
    m_threashold = threashold;
    m_gridArea = m_gridSize * m_gridSize;
    m_grid = new bool[m_gridArea];
    m_gridImage.create(m_gridSize, m_gridSize);
    m_gridTexture.loadFromImage(m_gridImage);
    reset();
}

void Grid::reset() {
    for (int i = 0; i < m_gridArea; i++) {
        m_grid[i] = (rand() % 100 + 1) / 100.f > m_threashold;
    }
}

void Grid::update() {
    bool *tempGrid = new bool[m_gridArea];
    std::copy(m_grid, m_grid + m_gridArea, tempGrid);

    for (int y = 0; y < m_gridSize; y++) {
        for (int x = 0; x < m_gridSize; x++) {
            int res = 0;

            int xx0 = x - 1;
            int yy0 = y - 1;
            int xx1 = x + 1;
            int yy1 = y + 1;

            if (x != 0)
                res += tempGrid[xx0 + m_gridSize * y] ? 1 : 0;

            if (y != 0)
                res += tempGrid[x + m_gridSize * yy0] ? 1 : 0;

            if (x != m_gridSize - 1)
                res += tempGrid[xx1 + m_gridSize * y] ? 1 : 0;

            if (y != m_gridSize - 1)
                res += tempGrid[x + m_gridSize * yy1] ? 1 : 0;

            if (x != 0 && y != 0)
                res += tempGrid[xx0 + m_gridSize * yy0] ? 1 : 0;

            if (x != m_gridSize - 1 && y != 0)
                res += tempGrid[xx1 + m_gridSize * yy0] ? 1 : 0;

            if (x != 0 && y != m_gridSize - 1)
                res += tempGrid[xx0 + m_gridSize * yy1] ? 1 : 0;

            if (x != m_gridSize - 1 && y != m_gridSize - 1)
                res += tempGrid[xx1 + m_gridSize * yy1] ? 1 : 0;

            if (!(tempGrid[x + m_gridSize * y] && (res == 3 || res == 2)))
                m_grid[x + m_gridSize * y] = false;
            if (!tempGrid[x + m_gridSize * y] && res == 3)
                m_grid[x + m_gridSize * y] = true;
        }
    }

    for (unsigned int x = 0; x < m_gridImage.getSize().x; x++) {
        for (unsigned int y = 0; y < m_gridImage.getSize().y; y++) {
            m_gridImage.setPixel(x, y, m_grid[y * m_gridImage.getSize().x + x] ? sf::Color::White : sf::Color::Black);
        }
    }
    m_gridTexture.update(m_gridImage);
}

sf::Texture &Grid::getTexture() { return m_gridTexture; }

sf::Image &Grid::getImage() { return m_gridImage; }